diff --git a/examples/example_convert.c b/examples/example_convert.c
index 7c0470d..7161eed 100644
--- a/examples/example_convert.c
+++ b/examples/example_convert.c
@@ -83,6 +83,22 @@ int main(int argc, char *argv[]) {
 
   //verify round trip conversion from RGB to NV12 and back to RGB
   img_t *nv12_image_round_trip = img_convert_rgb_to_nv12(image);
+  //print yuv values at center pixel
+  if (!nv12_image_round_trip) {
+    fprintf(stderr, "Failed to convert RGB to NV12 for round trip test\n");
+  } else {
+    uint8_t *y_plane = nv12_image_round_trip->planes[0];
+    uint8_t *uv_plane = nv12_image_round_trip->planes[1];
+    int center_index = (nv12_image_round_trip->height / 2) * nv12_image_round_trip->stride[0] + (nv12_image_round_trip->width / 2);
+    uint8_t Y = y_plane[center_index];
+    int center_uv_index = (nv12_image_round_trip->height / 4) * (nv12_image_round_trip->stride[1]) + (nv12_image_round_trip->width / 2);
+    uint8_t U = uv_plane[center_uv_index ]; // U value for the center pixel
+    uint8_t V = uv_plane[center_uv_index + 1];
+    printf("Center pixel in NV12: Y=%u U=%u V=%u\n", Y, U, V);
+    printf("Successfully converted RGB to NV12 for round trip test with dimensions %d x %d\n",
+           nv12_image_round_trip->width, nv12_image_round_trip->height);
+  }
+
   img_t *rgb_image_round_trip = img_convert_nv12_to_rgb(nv12_image_round_trip);
   if (!rgb_image_round_trip) {
     fprintf(stderr, "Failed to convert NV12 back to RGB\n");
@@ -91,10 +107,52 @@ int main(int argc, char *argv[]) {
     img_save_bmp("output_nv12_to_rgb.bmp", rgb_image_round_trip);
     printf("Successfully converted NV12 back to RGB with dimensions %d x %d\n",
            rgb_image_round_trip->width, rgb_image_round_trip->height);
+    //print rgb at center pixel
+    pixel_t center_pixel;
+    img_get_pixel(rgb_image_round_trip, rgb_image_round_trip->width / 2, rgb_image_round_trip->height / 2, &center_pixel);
+    printf("Center pixel in round trip NV12->RGB: R=%u G=%u B=%u\n", center_pixel.r, center_pixel.g, center_pixel.b);
     img_destroy(rgb_image_round_trip);
     rgb_image_round_trip = NULL;
   }
 
+  //Verify round trip conversion from RGB to YUV420P and back to RGB
+  img_t *yuv420p_image_round_trip = img_convert_rgb_to_yuv420p(image);
+  //print yuv values at center pixel
+  if (!yuv420p_image_round_trip) {
+    fprintf(stderr, "Failed to convert RGB to YUV420P for round trip test\n");
+  } else {
+    uint8_t *y_plane = yuv420p_image_round_trip->planes[0];
+    uint8_t *u_plane = yuv420p_image_round_trip->planes[1];
+    uint8_t *v_plane = yuv420p_image_round_trip->planes[2];
+    int center_index = (yuv420p_image_round_trip->height / 2) * yuv420p_image_round_trip->stride[0] + (yuv420p_image_round_trip->width / 2);
+    uint8_t Y = y_plane[center_index];
+    uint8_t U = u_plane[(yuv420p_image_round_trip->height / 2) * (yuv420p_image_round_trip->stride[1]) + (yuv420p_image_round_trip->width / 2)];
+    uint8_t V = v_plane[(yuv420p_image_round_trip->height / 2) * (yuv420p_image_round_trip->stride[2]) + (yuv420p_image_round_trip->width / 2)];
+    printf("Center pixel in YUV420P: Y=%u U=%u V=%u\n", Y, U, V);
+    printf("Successfully converted RGB to YUV420P for round trip test with dimensions %d x %d\n",
+           yuv420p_image_round_trip->width, yuv420p_image_round_trip->height);
+  }
+  img_t *rgb_image_round_trip_yuv = img_convert_yuv420p_to_rgb(yuv420p_image_round_trip);
+
+  if (!rgb_image_round_trip_yuv) {
+    fprintf(stderr, "Failed to convert YUV420P back to RGB\n");
+  } else {
+    img_toggle_rgb_bgr(rgb_image_round_trip_yuv); // Toggle back to BGR for correct color display in BMP
+    img_save_bmp("output_yuv420p_to_rgb.bmp", rgb_image_round_trip_yuv);
+    printf("Successfully converted YUV420P back to RGB with dimensions %d x %d\n",
+           rgb_image_round_trip_yuv->width, rgb_image_round_trip_yuv->height);
+    //print rgb at center pixel
+    pixel_t center_pixel;
+    img_get_pixel(rgb_image_round_trip_yuv, rgb_image_round_trip_yuv->width / 2, rgb_image_round_trip_yuv->height / 2, &center_pixel);
+    printf("Center pixel in round trip YUV420P->RGB: R=%u G=%u B=%u\n", center_pixel.r, center_pixel.g, center_pixel.b);
+    img_destroy(rgb_image_round_trip_yuv);
+    rgb_image_round_trip_yuv = NULL;
+  }
+
+  img_destroy(image);
+  image = NULL;
+  img_destroy(yuv420p_image_round_trip);
+  yuv420p_image_round_trip = NULL;
   img_destroy(nv12_image_round_trip);
   nv12_image_round_trip = NULL;
   img_destroy(copy);
diff --git a/include/img/img_convert.h b/include/img/img_convert.h
index 13e85fa..72d5432 100644
--- a/include/img/img_convert.h
+++ b/include/img/img_convert.h
@@ -15,4 +15,6 @@ int img_convert_rgb_to_yuv444(
 int img_convert_yuv444_to_rgb(img_t *img); // inplace conversion
 img_t *img_convert_rgb_to_nv12(const img_t *img);
 img_t *img_convert_nv12_to_rgb(const img_t *img);
+img_t *img_convert_rgb_to_yuv420p(const img_t *img);
+img_t *img_convert_yuv420p_to_rgb(const img_t *img);
 #endif
diff --git a/include/img/img_format.h b/include/img/img_format.h
index f3adf31..790f9a3 100644
--- a/include/img/img_format.h
+++ b/include/img/img_format.h
@@ -12,7 +12,8 @@ typedef enum {
   IMG_FMT_BGR24,
   IMG_FMT_RGBA32,
   IMG_FMT_YUV444,
-  IMG_FMT_NV12
+  IMG_FMT_NV12,
+  IMG_FMT_YUV420P
 } img_format_t;
 
 size_t img_format_bytes_per_pixel(img_format_t format);
diff --git a/src/core/img.c b/src/core/img.c
index b1c50fc..3f6cd3c 100644
--- a/src/core/img.c
+++ b/src/core/img.c
@@ -51,6 +51,62 @@ static img_t *img_create_nv12(uint32_t width, uint32_t height) {
   return img;
 }
 
+/*
+* Function to create a yuv420p image of width x height.
+ * return NULL on failure.
+ * return pointer to img_t on success.
+ */
+static img_t *img_create_yuv420p(uint32_t width, uint32_t height) {
+  if (width > SIZE_MAX || height > SIZE_MAX)
+    return NULL;
+
+  img_t *img = (img_t *)malloc(sizeof(img_t));
+  if (!img)
+    return NULL;
+  img->width = width;
+  img->height = height;
+  img->num_planes = 3; // YUV420P has 3 planes;
+  img->format = IMG_FMT_YUV420P;
+  img->stride[0] = width;
+  img->stride[1] = width / 2;
+  img->stride[2] = width / 2; // since YUV420P has 3 planes.
+
+  size_t buffer_size_Y_plane = width * height; // for Y plane.
+
+  // plane 0 will act as Y plane
+  img->planes[0] = malloc(buffer_size_Y_plane);
+  if (!img->planes[0]) {
+    free(img);
+    return NULL;
+  }
+
+  // plane 1 will act as U plane
+  size_t buffer_size_u_plane =
+      buffer_size_Y_plane / 4; // since U plane is a quarter of Y plane.
+  img->planes[1] = malloc(buffer_size_u_plane);
+  if (!img->planes[1]) {
+    free(img->planes[0]);
+    free(img);
+    return NULL;
+  }
+
+  // plane 2 will act as V plane
+  size_t buffer_size_v_plane =
+      buffer_size_Y_plane / 4; // since V plane is a quarter of Y plane.
+  img->planes[2] = malloc(buffer_size_v_plane);
+  if (!img->planes[2]) {
+    free(img->planes[0]);
+    free(img->planes[1]);
+    free(img);
+    return NULL;
+  }
+
+  memset(img->planes[0], 0, buffer_size_Y_plane);  // set 0 as initial value
+  memset(img->planes[1], 0, buffer_size_u_plane); // set 0 as initial value
+  memset(img->planes[2], 0, buffer_size_v_plane); // set 0 as initial value
+  return img;
+}
+
 /* Function to create an image of width x height.
  * return NULL on failure.
  * return pointer to img_t on success.
@@ -59,16 +115,31 @@ img_t *img_create(uint32_t width, uint32_t height, img_format_t format) {
   if (!width || !height)
     return NULL;
   bool is_planar = img_format_is_planar(format);
+
   if (is_planar) {
-    img_t *img_nv12 = img_create_nv12(width, height);
-    if (!img_nv12) {
-      fprintf(stderr, "Unable to create NV12 image\n");
+    switch (format) {
+    case IMG_FMT_NV12:
+      img_t *img_nv12 = img_create_nv12(width, height);
+      if (!img_nv12) {
+        fprintf(stderr, "Failed to create NV12 image\n");
+        return NULL;
+      }
+      return img_nv12;
+    case IMG_FMT_YUV420P:
+      img_t *img_yuv420p = img_create_yuv420p(width, height);
+      if (!img_yuv420p) {
+        fprintf(stderr, "Failed to create YUV420P image\n");
+        return NULL;
+      }
+      return img_yuv420p;
+      break;
+    default:
+      fprintf(stderr, "Unsupported planar format\n");
       return NULL;
     }
-    return img_nv12;
   }
-  size_t bpp = img_format_bytes_per_pixel(format);
 
+  size_t bpp = img_format_bytes_per_pixel(format);
   // Important checks.
   if (bpp == 0)
     return NULL;
@@ -78,6 +149,7 @@ img_t *img_create(uint32_t width, uint32_t height, img_format_t format) {
     return NULL;
 
   img_t *img = (img_t *)malloc(sizeof(img_t));
+  // printf("Allocated memory for img_t structure at %p\n", (void *)img);
   if (!img)
     return NULL;
 
@@ -89,7 +161,7 @@ img_t *img_create(uint32_t width, uint32_t height, img_format_t format) {
       width * bpp; // number of bytes in a row of pixel data, excluding padding
   size_t buffer_size = img->stride[0] * height;
   img->planes[0] = malloc(buffer_size);
-
+  // printf("Allocated memory for pixel data at %p\n", (void *)img->planes[0]);
   // Initialize other planes to NULL and strides to 0 for safety
   img->planes[1] = NULL;
   img->planes[2] = NULL;
diff --git a/src/core/img_convert.c b/src/core/img_convert.c
index 8178fa7..cbfba3f 100644
--- a/src/core/img_convert.c
+++ b/src/core/img_convert.c
@@ -4,6 +4,8 @@
 #include <stdio.h>
 #include <stdlib.h>
 
+#define CLAMP(x) (x < 0 ? 0 : (x > 255 ? 255 : (uint8_t)x))
+
 /* Convert RGB image to BGR
  * This function takes an image in RGB or BGR format and toggles it.
  * Parameters:  - img: A pointer to the image structure that contains the RGB or
@@ -288,13 +290,11 @@ img_t *img_convert_nv12_to_rgb(const img_t *img) {
   uint8_t *row_offset_uv;
   uint8_t *row_offset_rgb;
   int bpp = img_format_bytes_per_pixel(img_rgb->format);
-  for (int y = 0; y < img->height; y++)
-  {
+  for (int y = 0; y < img->height; y++) {
     row_offset_y = img->planes[0] + y * img->stride[0];
     row_offset_uv = img->planes[1] + (y / 2) * img->stride[1];
     row_offset_rgb = img_rgb->planes[0] + y * img_rgb->stride[0];
-    for (int x = 0; x < img->width; x++)
-    {
+    for (int x = 0; x < img->width; x++) {
       uint8_t *Y = row_offset_y + x;
 
       uint8_t *U = row_offset_uv + (x / 2) * 2;
@@ -313,7 +313,148 @@ img_t *img_convert_nv12_to_rgb(const img_t *img) {
       pixel[0] = R;
       pixel[1] = G;
       pixel[2] = B;
+    }
+  }
+  return img_rgb;
+}
+
+/*
+ * Convert RGB image to YUV420P
+ * This function converts an RGB image to YUV420P format.
+ * Parameters:
+ *   img: A pointer to the input RGB image
+ * Returns:
+ *   A pointer to the new YUV420P image on success, or NULL on failure
+ */
+img_t *img_convert_rgb_to_yuv420p(const img_t *img) {
+  if (!img) {
+    fprintf(stderr, "Input image is null\n");
+    return NULL;
+  }
+
+  if (img->format != IMG_FMT_RGB24) {
+    fprintf(stderr, "Input image format is %d, expected RGB24\n", img->format);
+    return NULL;
+  }
 
+  if (img->width % 2 != 0 || img->height % 2 != 0) {
+    fprintf(stderr, "Input image dimensions must be even for YUV420p format\n");
+    return NULL;
+  }
+
+  img_t *img_yuv420p = img_create(img->width, img->height, IMG_FMT_YUV420P);
+  if (!img_yuv420p) {
+    fprintf(stderr, "Failed to create YUV420P image\n");
+    return NULL;
+  }
+
+  int bpp = img_format_bytes_per_pixel(img->format);
+  uint8_t *row_offset_rgb;
+  uint8_t *row_offset_y;
+  uint8_t *row_offset_u;
+  uint8_t *row_offset_v;
+  uint8_t *pixel_rgb;
+  uint8_t *pixel_y;
+
+  for (int y = 0; y < img->height; y++) {
+    row_offset_rgb = img->planes[0] + y * img->stride[0];
+    row_offset_y = img_yuv420p->planes[0] + y * img_yuv420p->stride[0];
+    for (int x = 0; x < img->width; x++) {
+      pixel_rgb = row_offset_rgb + x * bpp;
+      pixel_y = row_offset_y + x;
+      pixel_y[0] = (uint8_t)(0.299 * pixel_rgb[0] + 0.587 * pixel_rgb[1] +
+                             0.114 * pixel_rgb[2]);
+    }
+  }
+
+  for (int y = 0; y < img->height; y += 2) {
+    row_offset_rgb = img->planes[0] + y * img->stride[0];
+    row_offset_u = img_yuv420p->planes[1] + (y / 2) * img_yuv420p->stride[1];
+    row_offset_v = img_yuv420p->planes[2] + (y / 2) * img_yuv420p->stride[2];
+    for (int x = 0; x < img->width; x += 2) {
+      uint8_t *pixel_u = row_offset_u + (x / 2);
+      uint8_t *pixel_v = row_offset_v + (x / 2);
+
+      uint8_t *pixel_tl = row_offset_rgb + x * bpp;
+      uint8_t *pixel_tr = row_offset_rgb + (x + 1) * bpp;
+      uint8_t *pixel_bl = row_offset_rgb + img->stride[0] + x * bpp;
+      uint8_t *pixel_br = row_offset_rgb + img->stride[0] + (x + 1) * bpp;
+
+      uint16_t r_sum = pixel_tl[0] + pixel_tr[0] + pixel_bl[0] + pixel_br[0];
+      uint16_t g_sum = pixel_tl[1] + pixel_tr[1] + pixel_bl[1] + pixel_br[1];
+      uint16_t b_sum = pixel_tl[2] + pixel_tr[2] + pixel_bl[2] + pixel_br[2];
+
+      uint8_t r_avg = (uint8_t)(r_sum / 4);
+      uint8_t g_avg = (uint8_t)(g_sum / 4);
+      uint8_t b_avg = (uint8_t)(b_sum / 4);
+
+      // Average the U and V values for the 2x2 block of pixels
+      uint8_t U = (uint8_t)(-0.169 * r_avg - 0.331 * g_avg + 0.5 * b_avg + 128);
+      uint8_t V = (uint8_t)(0.5 * r_avg - 0.419 * g_avg - 0.081 * b_avg + 128);
+
+      pixel_u[0] = U;
+      pixel_v[0] = V;
+    }
+  }
+  return img_yuv420p;
+}
+
+/*
+ * Convert YUV420P image to RGB
+ * This function converts a YUV420P image back to RGB format.
+ * Parameters:
+ *   img: A pointer to the input YUV420P image
+ * Returns:
+ *   A pointer to the new RGB image on success, or NULL on failure
+ */
+img_t *img_convert_yuv420p_to_rgb(const img_t *img) {
+  if (!img) {
+    fprintf(stderr, "Input image is null\n");
+    return NULL;
+  }
+
+  if (img->format != IMG_FMT_YUV420P) {
+    fprintf(stderr, "Input image format is %d, expected YUV420P\n",
+            img->format);
+    return NULL;
+  }
+
+  img_t *img_rgb = img_create(img->width, img->height, IMG_FMT_RGB24);
+
+  if (!img_rgb) {
+    fprintf(stderr, "Failed to create RGB image\n");
+    return NULL;
+  }
+
+  uint8_t *row_offset_y;
+  uint8_t *row_offset_u;
+  uint8_t *row_offset_v;
+  uint8_t *row_offset_rgb;
+  int bpp = img_format_bytes_per_pixel(img_rgb->format);
+
+  for (int y = 0; y < img->height; y++) {
+    row_offset_y = img->planes[0] + y * img->stride[0];
+    row_offset_u = img->planes[1] + (y / 2) * img->stride[1];
+    row_offset_v = img->planes[2] + (y / 2) * img->stride[2];
+    row_offset_rgb = img_rgb->planes[0] + y * img_rgb->stride[0];
+    for (int x = 0; x < img->width; x++) {
+      uint8_t *Y = row_offset_y + x;
+      uint8_t *U = row_offset_u + x / 2;
+      uint8_t *V = row_offset_v + x / 2;
+
+      double r_val = *Y + 1.402 * (*V - 128);
+      double g_val = *Y - 0.344 * (*U - 128) - 0.714 * (*V - 128);
+      double b_val = *Y + 1.772 * (*U - 128);
+
+      uint8_t *pixel = row_offset_rgb + x * bpp;
+
+      uint8_t R = CLAMP(r_val);
+      uint8_t G = CLAMP(g_val);
+      uint8_t B = CLAMP(b_val);
+
+      pixel[0] = R;
+      pixel[1] = G;
+      pixel[2] = B;
     }
   }
   return img_rgb;
diff --git a/src/core/img_format.c b/src/core/img_format.c
index 333eb9a..a07d7b0 100644
--- a/src/core/img_format.c
+++ b/src/core/img_format.c
@@ -24,6 +24,7 @@ size_t img_format_bytes_per_pixel(img_format_t format) {
 bool img_format_is_planar(img_format_t format) {
   switch (format) {
   case IMG_FMT_NV12:
+  case IMG_FMT_YUV420P:
     return true;
   default:
     return false;
diff --git a/tests/test_img_convert.c b/tests/test_img_convert.c
index 48aa1d2..09775d2 100644
--- a/tests/test_img_convert.c
+++ b/tests/test_img_convert.c
@@ -296,6 +296,56 @@ void test_img_convert_nv12_to_rgb() {
   img_destroy(nv12_img);
 }
 
+void test_img_convert_yuv420p_to_rgb_null() {
+  TEST_ASSERT_NULL(img_convert_yuv420p_to_rgb(NULL));
+}
+
+void test_img_convert_yuv420p_to_rgb() {
+  img_t *yuv420p_img = img_create(2, 2, IMG_FMT_YUV420P);
+  TEST_ASSERT_NOT_NULL(yuv420p_img);
+  // Manually set Y plane values
+  uint8_t *y_plane = yuv420p_img->planes[0];
+  y_plane[0] = 76; // Y value for Red
+  y_plane[1] = 150; // Y value for Green
+  y_plane[2] = 29; // Y value for Blue
+  y_plane[3] = 225; // Y value for Yellow
+  // Manually set U plane values (one U value for the entire 2x2 block)
+  uint8_t *u_plane = yuv420p_img->planes[1];
+  u_plane[0] = 96; // averaged U for the block
+  // Manually set V plane values (one V value for the entire 2x2 block)
+  uint8_t *v_plane = yuv420p_img->planes[2];
+  v_plane[0] = 133; // averaged V for the block
+  img_t *rgb_img = img_convert_yuv420p_to_rgb(yuv420p_img);
+  TEST_ASSERT_NOT_NULL(rgb_img);
+  TEST_ASSERT_EQUAL(IMG_FMT_RGB24, rgb_img->format);
+  TEST_ASSERT_EQUAL(yuv420p_img->width, rgb_img->width);
+  TEST_ASSERT_EQUAL(yuv420p_img->height, rgb_img->height);
+  // verify RGB values (should be close to original Red, Green, Blue, Yellow)
+  pixel_t out;
+  img_get_pixel(rgb_img, 0, 0, &out);
+  TEST_ASSERT_INT_WITHIN(5, 83, out.r); // Red remains Red
+  TEST_ASSERT_INT_WITHIN(5, 83, out.g);   // Green remains Green
+  TEST_ASSERT_INT_WITHIN(5, 19, out.b);   // Blue remains Blue
+
+  img_get_pixel(rgb_img, 1, 0, &out);
+  TEST_ASSERT_INT_WITHIN(5, 157, out.r);   // Red remains Red
+  TEST_ASSERT_INT_WITHIN(5, 157, out.g); // Green remains Green
+  TEST_ASSERT_INT_WITHIN(5, 94, out.b);   // Blue remains Blue
+
+  img_get_pixel(rgb_img, 0, 1, &out);
+  TEST_ASSERT_INT_WITHIN(5, 36, out.r);   // Red remains Red
+  TEST_ASSERT_INT_WITHIN(5, 36, out.g);   // Green remains Green
+  TEST_ASSERT_INT_WITHIN(5, 0, out.b); // Blue remains Blue
+
+  img_get_pixel(rgb_img, 1, 1, &out);
+  TEST_ASSERT_INT_WITHIN(5, 232, out.r); // Red remains Red
+  TEST_ASSERT_INT_WITHIN(5, 232, out.g); // Green remains Green
+  TEST_ASSERT_INT_WITHIN(5, 169, out.b);   // Blue remains Blue
+  
+  img_destroy(rgb_img);
+  img_destroy(yuv420p_img);
+}
+
 int main() {
   UNITY_BEGIN();
   RUN_TEST(test_img_toggle_rgb_bgr_null);
@@ -316,6 +366,7 @@ int main() {
   RUN_TEST(test_img_convert_rgb_to_nv12);
   RUN_TEST(test_img_convert_nv12_to_rgb_null);
   RUN_TEST(test_img_convert_nv12_to_rgb);
-
+  RUN_TEST(test_img_convert_yuv420p_to_rgb_null);
+  RUN_TEST(test_img_convert_yuv420p_to_rgb);
   return UNITY_END();
 }
\ No newline at end of file
